src/ast/semantic_analyser.cpp

```diff
@@ -607,11 +607,15 @@ void SemanticAnalyser::visit(Call &call)
       if (call.vargs->size() == 2)
       {
         auto &size_arg = *call.vargs->at(1);
-        long value = static_cast<Integer &>(size_arg).n;
-        if (value < 0)
+        auto integer = dynamic_cast<Integer *>(&size_arg);
+        if (integer)
```

Comment:
I think this check was wrong to begin with. We should only be downcasting and checking the value after we know it's a literal, right? Otherwise we could be reading junk. So in `is_final_pass()` and after the `check_arg()`. Same for the `buf` case. I believe that would solve this static_cast issue as well.

Reply:
That makes sense. But I found in the case of buf(), the length (buffer_size) is used in other places, and we cannot check the length in the final pass.. So I decided to use dynamic_cast here. 735b944 please let me know if there is a better way.

Reply:
Can't buf() just use `max_buffer_size` until the final pass? Then on the final pass it can do all the error checks and calculations.

Reply:
I first tried it and failed. Then I realized that AssignMap did not update the map size. I fixed it and now I think it's OK: bfded77

Reply:
Is update_assign_map_type() fixing a separate issue? I'm not seeing why it's related.

Reply:
without it, https://github.com/iovisor/bpftrace/blob/7ca62cb0aca9d6125d8a96fdc0238f7e07354097/src/ast/semantic_analyser.cpp#L2067 is failed because it thinks map size is `max_buffer_size`, but `expr_size` become smaller in the final pass.